lib/ExecutionEngine/ExecutionEngine.cpp

```diff
@@ -93,16 +110,30 @@ void ExecutionEngine::runBatch(size_t iterations,
   }
 }
 
-void ExecutionEngine::updateInputsAndRunNetwork(llvm::ArrayRef<Variable *> vars,
+void ExecutionEngine::updateInputsAndRunNetwork(llvm::ArrayRef<Storage *> vars,
                                                 llvm::ArrayRef<Tensor *> inputs,
                                                 size_t sampleIdx) {
+
+  llvm::SmallVector<Placeholder *, 8> placeholders;
+  llvm::SmallVector<Tensor *, 8> tensors;
+
   // Update the input variables.
   for (int i = 0, e = vars.size(); i < e; i++) {
-    loadValueFromTensorSlice(vars[i], inputs[i], sampleIdx);
+    // Note: When we finish the migration to Placeholder this code needs to be
+    // removed because it's not threadsafe. See issue #1334.
+    if (Variable *V = llvm::dyn_cast<Variable>(vars[i])) {
+      loadValueFromTensorSlice(V, inputs[i], sampleIdx);
+      continue;
+    }
+
+    // This is a placeholder that we need to make concrete during the execution
+    // of the program.
+    placeholders.push_back(cast<Placeholder>(vars[i]));
+    tensors.push_back(inputs[i]);
```

Comment:
Don't we need to get slices for the placeholders, like we currently do for Variables?

Reply:
Yes. You are right. I missed that. I am not sure how to handle this right now. I think that the runBatch interface is not great. Maybe we should get rid of it and turn it into a utility that the unit tests use. 

I'll try to think of a way to fix this. Maybe I'll create a tensor view for the lifetime of the function.